### PR TITLE
Build toolchain with Github Actions

### PR DESCRIPTION
**Pull request by krystalgamer**
This PR makes it possible to automatically build the toolchain which produces an artifact with roughly ~40mb.

Currently on supports linux.

### CODE REVIEW

**Review by sajattack — APPROVED** [review]

**Comment by carstene1ns** [comment]
Might not have seen it, but there is already an approach in the 'toolchain-only' branch? https://github.com/pspdev/psptoolchain/blob/toolchain-only/.github/workflows/compilation.yml

**Comment by krystalgamer** [comment]
> Might not have seen it, but there is already an approach in the 'toolchain-only' branch? https://github.com/pspdev/psptoolchain/blob/toolchain-only/.github/workflows/compilation.yml

Didn't know you already had one :/

Comparing both

* Mine uses the repo scripts while yours has the dependencies installed manually
* Yours has support for Mac
* Mine uploads the artifacts

Either way I'd be happy if either is merged, having automatic available builds would imho make it much easier to setup dev environments.


Having this merged has a huge potential not only for people who just want the binaries but also for the capability of building PSP homebrew with github actions via something like  `setup-psptoolchain`.

**Comment by carstene1ns** [comment]
For building external things with github actions it should not be needed to built the toolchain each run. There is a docker image which can be easily used.

**Comment by krystalgamer** [comment]
> For building external things with github actions it should not be needed to built the toolchain each run.

That's not what I said. `setup-toolchain` gets the latest psptoolchain already compiled from the GH Releases and sets up the environment variables. 
The whole setting up takes 5s: https://github.com/krystalgamer/UmdMount/runs/1943104739

Compiling the whole toolchain for each run is nonsense and I'm not sure where you got that idea.


> There is a docker image which can be easily used.

I disagree, first the docker image is out-of-date: https://hub.docker.com/r/pspdev/psptoolchain
Pulling the image would take much longer and wouldn't beat a one-liner: https://github.com/krystalgamer/UmdMount/blob/master/.github/workflows/c-cpp.yml#L15

Using docker in GH actions is counter-intuitive, the provided images include a developer environment with almost all dependencies. Also you can take the output of my workflow and dockerize it, but taking the toolchain off the docker image would be annoying.

**Comment by fjtrujy** [comment]
Hello guys,

I see here 2 different things:
1. @krystalgamer you just did a workflow for generating an artifact (just Ubuntu) for saving time developers compiling locally (which is not always true & working)
2. @carstene1ns directions is something more scalable and pointing to a full CI/CD integration. Having separate repositories, each of them generating a docker layer, and finally offering as well a `preinstalled` toolchain (in a new repo called `pspdev`)

So, I think that this PR, will not make sense as soon as the @carstene1ns are done.

Thanks

**Comment by sharkwouter** [comment]
You may want to make this manually toggle-able as well or add a cron job actually. Since this will need to be updated whenever a change is made to any of the involved repos.

**Comment by fjtrujy** [comment]
@sharkwouter the @carstene1ns direction is exactly this, to automatically release a new version every time that a commit is done in the other repositories, that is why it requires to have a better repo-separation.

**Comment by krystalgamer** [comment]
> You may want to make this manually toggle-able as well or add a cron job actually. Since this will need to be updated whenever a change is made to any of the involved repos.

By default they're manually toggle-able. Just press "Re-run jobs". GH doesn't support cron jobs but from what @fjtrujy said yesterday on Discord repos can notify each other.


> 1. @krystalgamer you just did a workflow for generating an artifact (just Ubuntu) for saving time developers compiling locally (which is not always true & working)

I didn't do it to save time, my PC does it much faster. The end goal was to have a distributable binary release of the toolchain always up2date.
Additionally it could also be used to build homebrew directly on GH.

> 2. @carstene1ns directions is something more scalable and pointing to a full CI/CD integration. Having separate repositories, each of them generating a docker layer, and finally offering as well a `preinstalled` toolchain (in a new repo called `pspdev`)

A full CI/CD integration doesn't require docker and imho it's shouldn't. Docker is only decent on Linux any other OS is just agony.
The only difference is that mine publishes a tarball while `psp-toolchain` pushes to docker-hub. 
Also not sure what you mean with "more scalable", both pipelines can handle the toolchain just fine.


> So, I think that this PR, will not make sense as soon as the @carstene1ns are done.

It's been sitting dormant for 8 months with no updates. This can at least be a temporary workflow while yours is being worked on.

**Comment by fjtrujy** [comment]
> I didn't do it to save time, my PC does it much faster. The end goal was to have a distributable binary release of the toolchain always up2date.
As I said before, we also have it in `ps2dev` https://github.com/ps2dev/ps2dev/releases/tag/v1.2.0
and I can assure you that is not always working (don't ask me why), actually MacOS one, looks to work better than Linux.

> A full CI/CD integration doesn't require docker and imho it's shouldn't. Docker is only decent on Linux any other OS is just agony.
Yes, it is MANDATORY. So imagine, a PR done in `pspsdk`, this PR is merged because we "trust" in the changes. Now you have:
1. Manually trigger a rebuild of the artifacts
2. You can't assure before applying the commit to `pspsdk` that this commit is not breaking your CI
3. So in order to check if the commit is working before get merged in `pspsdk` you need to implement CI/CD in the `pspsdk` repo. How you do it? is here where it can be done out of the box using a docker image.

As I said @carstene1ns goal is also to auto-distribute an artifact (as you're doing), but it is going even farther.
If you go to the next link, you will see better what we meant: 
https://github.com/pspdev/pspdev-docker/issues/12 

Thanks

File: .github/workflows/c-cpp.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      run: sudo dpkg-reconfigure --frontend=noninteractive dash
+    - name: Compile
+      run: sudo ./toolchain-sudo.sh
+    - name: Pack to tar.xz
+      run: tar -cf - /usr/local/pspdev | xz --verbose -9 -T0 -c - > toolchain.tar.xz
+    - name: Upload Artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: psptoolchain.tar.xz
+        path: toolchain.tar.xz
```
Comment (Margen67) [inline .github/workflows/c-cpp.yml:30]:
```suggestion
        path: toolchain.tar.xz
        if-no-files-found: error
```

**Comment by wally4000** [comment]
Has been fulfilled by fjtrujy using an alternative method, closing to prevent merge conflicts.

Thanks for your contribution though krystalgamer :)

**Pull request closed without merging**